metricbeat/mb/parse/url.go

```diff
@@ -138,6 +139,7 @@ func NewHostDataFromURL(u *url.URL) mb.HostData {
 // Values from the rawHost take precedence over the defaults.
 func ParseURL(rawHost, scheme, user, pass, path, query string) (mb.HostData, error) {
```

Comment:
Maybe I'm missing something, but what if we introduced a new field called something like `Transport` in the `mb.HostData` struct? The field could be an enum of `tcp`, `unix` or `npipe`.

Then, in this `parse.ParseURL` function here, we could determine the transport from the `rawHost` value. After determining the transport, we massage the `rawHost` value such that it does not contain `npipe` or `unix`, before we pass it into `getURL` — essentially, we maintain the same contract for `getURL` as before.

Then, we change the signature of the `NewHostDataFromURL` function to take in the transport parameter as well, and use it when constructing the `mb.HostData` struct that is returned. I understand that this means changing the calls to this function in the mongodb and postgres modules as well but that should be a matter of just adding the transport argument as `tcp`.

Finally, in `helper.newHTTPFromConfig`, we can pass the value of this `transport` field as an argument to the new `makeDialier` function.

This way we don't have to rely on the the special `__path` query parameter.

WDYT?

Reply:
A small variation could also be to keep the `NewHostDataFromURL` function signature the same and create a new function like `NewHostDataFromURLWithTransport`. The implementation of `NewHostDataFromURL` would be to simply call `NewHostDataFromURLWithTransport` with the `tcp` transport enum. That way you don't need to touch the `mongodb` or `postgresql` Metricbeat modules code.

Reply:
I think its a better solution than changing everything. ^

Reply:
@ycombinator I'll also need to add a `Path` field to the `HostData` to make sure **path to the socket** and the actual HTTP path is kept separated. Because the real issue in the current code base is that if you want to either use a UNIX socket or a npipe, any **HTTP** path will be ignored simply because when you parse a `unix::///tmp/fb.sock` with Go url parse the "Path" field will point to the `/tmp/fb.sock` and will make the [ code](https://github.com/elastic/beats/blob/eff54c3069a63fe5c91b78c4edf15a6505c56759/metricbeat/mb/parse/url.go#L212) in the `getURL` ignore any added path or subpath.

Reply:
Actually looking even more at the code in `getURL` I would be inclined to think it's broken for following use case.

Scenario: I have beats monitoring turned on, the beats is behind a proxy and you now have base path. 

So if you set the URL in the beat module toi`http://myhost/basepath/` we should append the monitoring path to the base path and actually call `http://myhost/basepath/states`